### PR TITLE
[x/programs] Serialize params on the Rust side

### PR DESCRIPTION
**Pull request by iFrostizz**
Closes https://github.com/ava-labs/hypersdk/issues/927

### CODE REVIEW

File: x/programs/cmd/simulator/cmd/plan.go
Comment (iFrostizz) [inline x/programs/cmd/simulator/cmd/plan.go:365]: we were not using this

File: x/programs/cmd/simulator/src/lib.rs
Comment (iFrostizz) [inline x/programs/cmd/simulator/src/lib.rs:332]: We never needed to deserialize a `Param`, so I removed the derive attribute as well as the test

File: x/programs/cmd/simulator/src/lib.rs
Comment (iFrostizz) [inline x/programs/cmd/simulator/src/lib.rs:8]: The `serde_with` dependency will be dropped once we merge https://github.com/ava-labs/hypersdk/pull/935

**Review by dboehm-avalabs — DISMISSED** [review]

File: x/programs/cmd/simulator/src/lib.rs
Comment (iFrostizz) [reply x/programs/cmd/simulator/src/lib.rs:8]: edit: probably after manually implementing `Serialize` on the steps once we get rid of plans.

File: x/programs/cmd/simulator/cmd/plan.go
Comment (richardpringle) [inline x/programs/cmd/simulator/cmd/plan.go:208]: Why not just use the index instead of the string, the `step_` is redundant here. You could probably also use a sparse `[]id.ID` and fill in a bunch `nil` for all the other outputs. None of this needs to be performant.

File: x/programs/cmd/simulator/cmd/plan.go
Comment (richardpringle) [inline x/programs/cmd/simulator/cmd/plan.go:258]: remind me again why it's `param[1:]`?

File: x/programs/cmd/simulator/cmd/plan.go
Comment (richardpringle) [reply x/programs/cmd/simulator/cmd/plan.go:369]: I'm sure there's more unused code lurking too

File: x/programs/cmd/simulator/src/lib.rs
Comment (richardpringle) [inline x/programs/cmd/simulator/src/lib.rs:112]: Haha this is definitely a big *nit*, but I would do this:
```suggestion
        let mut serializer = serializer.serialize_struct("param", 2)?;
```

File: x/programs/cmd/simulator/src/lib.rs
Comment (richardpringle) [inline x/programs/cmd/simulator/src/lib.rs:6]: Is there any real benefit to base64 encoding everything?

File: x/programs/cmd/simulator/src/lib.rs
Comment (richardpringle) [reply x/programs/cmd/simulator/src/lib.rs:139]: I don't quite understand why everything needs to be base64 encoded. If I had to guess, it's because it's easier to decode everything on the Go side, but you should just be able to *switch* on the `type` tag and only base64 decode where necessary. But IMO, it's an extra step that's unnecessary (and adds a little complexity). 

Also, I think it would just be beneficial to use an intermediate type here:
```rust
#[derive(Serialize)]
#[serde(rename_all = "lowercase", tag = "type", content = "value")]
enum StringParam {
    U64(String),
    String(String),
    Id(String),
    #[serde(untagged)]
    Key(Key),
}
```
If you want, you can define the type inside the `Serialize` implementation, do a `From` conversion, then just serialize the intermediate type. This should all get optimized away by the compiler, especially when the type is only used in one place. 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=755c35cea6729ce0da44df255dceeaed

**Review by richardpringle — COMMENTED** [review]
Leaving this as a comment review since I'll need to re-approve anyway when the merge conflicts are resolved

File: x/programs/cmd/simulator/src/lib.rs
Comment (iFrostizz) [reply x/programs/cmd/simulator/src/lib.rs:139]: It's Go making choices for us when dealing with the `[]byte` type, see https://pkg.go.dev/encoding/json#Marshal
> Array and slice values encode as JSON arrays, except that []byte encodes as a base64-encoded string, and a nil slice encodes as the null JSON value.
I found online that people were forking the package to strip out the base64 part but I don't think that we should do that.

File: x/programs/cmd/simulator/cmd/plan.go
Comment (iFrostizz) [reply x/programs/cmd/simulator/cmd/plan.go:208]: Replaced the type of the key of the map to `int`. Using a list would also have worked but would have required a bit more of complexity.

File: x/programs/cmd/simulator/cmd/plan.go
Comment (iFrostizz) [reply x/programs/cmd/simulator/cmd/plan.go:258]: That's because within all params, the id of the program that is called is passed as the first of them.
The first parameter always has a special treatment and it is defined in this function that you tagged

File: x/programs/cmd/simulator/src/lib.rs
Comment (iFrostizz) [reply x/programs/cmd/simulator/src/lib.rs:112]: 😨 
(replaced)

File: x/programs/cmd/simulator/src/lib.rs
Comment (iFrostizz) [reply x/programs/cmd/simulator/src/lib.rs:6]: https://github.com/ava-labs/hypersdk/issues/994

File: x/programs/cmd/simulator/Cargo.toml
Comment (iFrostizz) [inline x/programs/cmd/simulator/Cargo.toml:10]: I promised!

File: x/programs/cmd/simulator/src/lib.rs
Comment (richardpringle) [reply x/programs/cmd/simulator/src/lib.rs:125]: Why not just use `usize::to_string`. It's not fallible

File: x/programs/cmd/simulator/src/lib.rs
Comment (richardpringle) [inline x/programs/cmd/simulator/src/lib.rs:204]:
```suggestion
fn b64_encode<S>(text: &str, serializer: S) -> Result<S::Ok, S::Error>
```

Comment (richardpringle) [reply x/programs/cmd/simulator/src/lib.rs:204]: And while you're at it... I really would prefer:
```suggestion
fn base64_encode<S>(text: &str, serializer: S) -> Result<S::Ok, S::Error>
```
But this is a nit

File: x/programs/cmd/simulator/src/lib.rs
Comment (richardpringle) [inline x/programs/cmd/simulator/src/lib.rs:215]:
```suggestion
    <String as Deserialize>::deserialize(deserializer).and_then(|string| {
```

**Review by richardpringle — CHANGES_REQUESTED** [review]

File: x/programs/cmd/simulator/src/lib.rs
Comment (iFrostizz) [reply x/programs/cmd/simulator/src/lib.rs:125]: Because it still has the `step_` prefix, but it probably makes sense to drop it here too!

File: x/programs/cmd/simulator/src/lib.rs
Comment (iFrostizz) [reply x/programs/cmd/simulator/src/lib.rs:204]: Fixed!

File: x/programs/cmd/simulator/src/lib.rs
Comment (iFrostizz) [reply x/programs/cmd/simulator/src/lib.rs:215]: This is very nice! thanks

File: x/programs/cmd/simulator/src/lib.rs
Comment (iFrostizz) [reply x/programs/cmd/simulator/src/lib.rs:125]: I have implemented `From<&usize>` instead and applied your suggestions. That's better!

File: x/programs/cmd/simulator/src/id.rs
Comment (richardpringle) [inline x/programs/cmd/simulator/src/id.rs:16]: 👍

File: x/programs/cmd/simulator/src/lib.rs
Comment (richardpringle) [inline x/programs/cmd/simulator/src/lib.rs:131]: This is a little more complex than what I had suggested. Also, you really don't need the `step_`. In json, the `"type"` field will be `"id"`, right? that means that `step_` is redundant.

**Review by richardpringle — APPROVED** [review]
All the base-64 stuff is a little convoluted. This is a good step into cleaning things up, but I think there's still more to do.

**Pull request merged**
